contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,273 @@
+    }
+
+    /// @notice the rate per second for each address
+    function getRateLimitPerSecond(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].rateLimitPerSecond;
+    }
+
+    /// @notice the last time the buffer was used by each address
+    function getLastBufferUsedTime(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].lastBufferUsedTime;
+    }
+
+    /// @notice the cap of the buffer that can be used at once
+    function getBufferCap(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].bufferCap;
+    }
+
+    // ----------- Helper Methods -----------
+
+    function _updateAddress(address rateLimitedAddress, uint112 _rateLimitPerSecond, uint144 _bufferCap) internal {
+        uint112 oldRateLimitPerSecond = _rateLimitPerSecond;
+
+        RateLimitData storage rateLimitData = rateLimitPerAddress[rateLimitedAddress];
+
+        require(rateLimitData.lastBufferUsedTime != 0, "MultiRateLimited: rate limit address does not exist");
+        require(_rateLimitPerSecond <= MAX_RATE_LIMIT_PER_SECOND, "MultiRateLimited: rateLimitPerSecond too high");
+
+        rateLimitData.lastBufferUsedTime = block.timestamp.toUint32();
+        rateLimitData.bufferCap = _bufferCap;
+        rateLimitData.rateLimitPerSecond = _rateLimitPerSecond;
+        rateLimitData.bufferStored = _bufferCap;
+
+        emit IndividualRateLimitPerSecondUpdate(rateLimitedAddress, oldRateLimitPerSecond, _rateLimitPerSecond);
+    }
+
+    /// @param rateLimitedAddress the new address to add as a rateLimitedAddress
+    /// @param _rateLimitPerSecond the rate limit per second for this rateLimitedAddress
+    /// @param _bufferCap  the buffer cap for this rateLimitedAddress
+    function _addAddress(address rateLimitedAddress, uint112 _rateLimitPerSecond, uint144 _bufferCap) internal {
+        require(_bufferCap <= bufferCap, "MultiRateLimited: new buffercap too high");
+        require(rateLimitPerAddress[rateLimitedAddress].lastBufferUsedTime == 0, "MultiRateLimited: address already added");
+        require(_rateLimitPerSecond <= MAX_RATE_LIMIT_PER_SECOND, "MultiRateLimited: rateLimitPerSecond too high");
+
+        RateLimitData memory rateLimitData = RateLimitData({
+            lastBufferUsedTime: block.timestamp.toUint32(),
+            bufferCap: _bufferCap,
+            rateLimitPerSecond: _rateLimitPerSecond,
+            bufferStored: _bufferCap
+        });
+
+        rateLimitPerAddress[rateLimitedAddress] = rateLimitData;
+
+        emit IndividualRateLimitPerSecondUpdate(rateLimitedAddress, 0, _rateLimitPerSecond);
+    }
+
+    /// @notice the method that enforces the rate limit. Decreases buffer by "amount". 
+    /// If buffer is <= amount either
+    /// 1. Does a partial depletion by the amount remaining in the buffer or
```

Comment:
I really think we should disable partial depletions here. It's not a widely used feature, and disabling it closes off a whole branch of security analysis we won't have to do.

Reply:
If it doesn't allow partial depletion, then the comment should say that.
Actually, if this functionality doesn't exist at all, why is it a parameter being passed to MRL?